core/src/dsnp/reader_writer.rs

```diff
@@ -12,28 +13,40 @@ use anyhow::Result;
 
 /// DSNP compatible reader
 pub trait DsnpReader {
+	type Encryption: EncryptionBehavior;
 	/// reading public key from binary
 	fn read_public_key(data: &[u8]) -> Result<DsnpPublicKey>;
 	/// reading public graph from binary
 	fn read_public_graph(data: &[u8]) -> Result<DsnpInnerGraph>;
 	/// reading private graph from binary
-	fn read_private_graph(data: &[u8]) -> Result<PrivateGraphChunk>;
+	fn read_private_graph(
+		data: &[u8],
+		decryption_input: &<Self::Encryption as EncryptionBehavior>::DecryptionInput,
+	) -> Result<PrivateGraphChunk>;
 }
 
 /// DSNP compatible writer
 pub trait DsnpWriter {
+	type Encryption: EncryptionBehavior;
 	/// write public key to binary
 	fn write_public_key(key: &DsnpPublicKey) -> Result<Vec<u8>>;
 	/// write public graph to binary
 	fn write_public_graph(inner: &DsnpInnerGraph) -> Result<Vec<u8>>;
 	/// write private graph to binary
-	fn write_private_graph(graph: &PrivateGraphChunk) -> Result<Vec<u8>>;
+	fn write_private_graph(
+		graph: &PrivateGraphChunk,
+		encryption_input: &<Self::Encryption as EncryptionBehavior>::EncryptionInput,
+	) -> Result<Vec<u8>>;
 }
 
 /// A utility to read/write data from and to Frequency chain specific implementation of DSNP
 pub struct Frequency;
 
+/// implementing DsnpReader for Frequency
 impl DsnpReader for Frequency {
+	/// using SealBox for encryption
+	type Encryption = SealBox;
```

Comment:
Is there a way to to assign the EncryptionBehavior implementation on struct level so we don't have to assign it one DsnpReader and DsnpWriter separately!?